### ISSUE TITLE
[docs] Dynamically generate TOC

### ISSUE THREAD

**Issue by imphil** [lowRISC/opentitan#1419]
The page https://docs.opentitan.org/ has a TOC-like structure in it, as does https://docs.opentitan.org/doc/ug/. But the contents are manually edited and don't reflect all subsections. 

Can we auto-generate this content to keep it up-to-date and ensure that all content is actually linked?

**Comment by sjgitty** [comment]
For this one we should consider whether a true TOC is what we want.
It is *possibly* going to look good but sometimes TOCs look ungainly.
Below is what it would look like if we were to render all of the document
titles as the TOC entries. Clearly it is very deep (I'm sure we could employ
some .skip directives), but also the algorithm to generate this was not
obvious (I had to kluge it several times), so not sure what algorithm we
would want to use (perhaps there are standard Hugo things we could use
here). It also shows that we need to clean up a lot of our document titles
which are not self-explanatory. Thoughts?

* [Contributing code to the opentitan repository](CONTRIBUTING.md)                                  
* [OpenTitan](README.md)                                                                            
* [lowRISC Documentation Overview](doc/_index.md)                                                   
  * [Introduction to the OpenTitan Project](doc/project/_index.md)                                  
    * [Signoff Checklist](doc/project/checklist.md)                                                 
    * [Committers](doc/project/committers.md)                                                       
    * [Hardware Designs Dashboard](doc/project/hw_dashboard.md)                                     
    * [OpenTitan Hardware Development Stages](doc/project/hw_stages.md)                             
    * [OpenTitan RFC Process](doc/project/rfc_process.md)                                           
  * [Reference Manuals](doc/rm/_index.md)                                                           
    * [RISC-V Assembly Style Guide](doc/rm/asm_coding_style.md)                                     
    * [C and C++ Coding Style Guide](doc/rm/c_cpp_coding_style.md)                                  
    * [Hjson Usage and Style Guide](doc/rm/hjson_usage_style.md)                                    
    * [Markdown Usage and Style Guide](doc/rm/markdown_usage_style.md)                              
    * [Python Coding Style Guide](doc/rm/python_coding_style.md)                                    
    * [FPGA Reference Manual](doc/rm/ref_manual_fpga.md)                                            
    * [Top Generation Tool](doc/rm/topgen_tool.md)                                                  
    * [util/vendor.py: Vendor-in Components](doc/rm/vendor_in_tool.md)                              
    * [Verilog Coding Style Guide](doc/rm/verilog_coding_style.md)                                  
  * [OpenTitan Security Overview](doc/security/_index.md)                                           
  * [User Guides](doc/ug/_index.md)                                                                 
    * [Design Methodology within OpenTitan](doc/ug/design.md)                                       
    * [Directory Structure](doc/ug/directory_structure.md)                                          
    * [Design Verification Methodology within OpenTitan](doc/ug/dv_methodology.md)                  
    * [Get an FPGA Board](doc/ug/fpga_boards.md)                                                    
    * [Getting Started with an OpenTitan Hardware Design](doc/ug/getting_started_design.md)         
    * [Getting Started with an OpenTitan Design Verification](doc/ug/getting_started_dv.md)         
    * [Getting started on FPGAs](doc/ug/getting_started_fpga.md)                                    
    * [Getting Started](doc/ug/getting_started.md)                                                  
    * [Build Software](doc/ug/getting_started_sw.md)                                                
    * [Getting started with Verilator](doc/ug/getting_started_verilator.md)                         
    * [GitHub Notes](doc/ug/github_notes.md)                                                        
    * [Quickstart](doc/ug/quickstart.md)                                                            
    * [List of Top-Level Designs](doc/ug/system_list.md)                                            
    * [Work with hardware code in external repositories](doc/ug/vendor_hw.md)                       
* [Hardware Specifications](hw/_index.md)                                                           
  * [OpenTitan Assertions](hw/formal/README.md)                                                     
  * [RTL Linting](hw/lint/README.md)                                                                
  * [Top Earlgrey](hw/top_earlgrey/README.md)                                                       
  * [Third-Party/External Code](hw/vendor/README.md)                                                
    * [FOO DV Plan](hw/dv/doc/dv_plan_template.md)                                                  
    * [DV build and run flow](hw/dv/tools/README.md)                                                
      * [JTAG DPI module for OpenOCD remote_bitbang driver](hw/dv/dpi/jtagdpi/README.md)            
      * [ALERT Agent](hw/dv/sv/alert_esc_agent/README.md)                                           
      * [Common interfaces](hw/dv/sv/common_ifs/index.md)                                           
      * [CSR utilities](hw/dv/sv/csr_utils/README.md)                                               
      * [DV Library Classes](hw/dv/sv/dv_lib/README.md)                                             
      * [README.md](hw/dv/sv/dv_utils/README.md)                                                    
      * [I2C DV UVM Agent](hw/dv/sv/i2c_agent/README.md)                                            
      * [JTAG DV UVM Agent](hw/dv/sv/jtag_agent/README.md)                                          
      * [Mem_bkdr Interface](hw/dv/sv/mem_bkdr_if/README.md)                                        
      * [Memory Model](hw/dv/sv/mem_model/README.md)                                                
      * [README.md](hw/dv/sv/scoreboard/README.md)                                                  
      * [SPI UVM Agent](hw/dv/sv/spi_agent/README.md)                                               
      * [test_vectors_pkg.sv](hw/dv/sv/test_vectors/README.md)                                      
      * [TileLink UVM Agent](hw/dv/sv/tl_agent/README.md)                                           
      * [UART Agent](hw/dv/sv/uart_agent/README.md)                                                 
      * [USB20 UVM Agent](hw/dv/sv/usb20_agent/README.md)                                           
        * [Comportable IP Testbench Architecture](hw/dv/sv/cip_lib/doc/index.md)                    
    * [RISC-V Platform-Level Interrupt Controller](hw/ip/rv_plic/README.md)                         
      * [AES HWIP Technical Specification](hw/ip/aes/doc/_index.md)                                 
        * [AES Checklist](hw/ip/aes/doc/checklist.md)                                               
      * [AES Model](hw/ip/aes/model/README.md)                                                      
        * [AES DV Plan](hw/ip/aes/doc/dv_plan/index.md)                                             
      * [Alert Handler Technical Specification](hw/ip/alert_handler/doc/_index.md)                  
        * [Alert Handler Checklist](hw/ip/alert_handler/doc/checklist.md)                           
          * [ALERT_HANDLER DV Plan](hw/ip/alert_handler/doc/dv_plan/index.md)                       
      * [ENTROPY_SRC HWIP Technical Specification](hw/ip/entropy_src/doc/_index.md)                 
      * [Flash Controller HWIP Technical Specification](hw/ip/flash_ctrl/doc/_index.md)             
        * [FLASH_CTRL Checklist](hw/ip/flash_ctrl/doc/checklist.md)                                 
      * [GPIO HWIP Technical Specification](hw/ip/gpio/doc/_index.md)                               
        * [GPIO Checklist](hw/ip/gpio/doc/checklist.md)                                             
          * [GPIO DV Plan](hw/ip/gpio/doc/dv_plan/index.md)                                         
      * [HMAC HWIP Technical Specification](hw/ip/hmac/doc/_index.md)                               
        * [HMAC Checklist](hw/ip/hmac/doc/checklist.md)                                             
          * [HMAC DV Plan](hw/ip/hmac/doc/dv_plan/index.md)                                         
          * [Cryptoc library (SHA, SHA256 and HMAC)](hw/ip/hmac/dv/cryptoc_dpi/README.md)           
      * [I2C HWIP Technical Specification](hw/ip/i2c/doc/_index.md)                                 
        * [NMI Generator Technical Specification](hw/ip/nmi_gen/doc/index.md)                       
          * [I2C DV Plan](hw/ip/i2c/doc/dv_plan/index.md)                                           
      * [Padctrl Technical Specification](hw/ip/padctrl/doc/_index.md)                              
        * [Padctrl Checklist](hw/ip/padctrl/doc/checklist.md)                                       
      * [Pinmux Technical Specification](hw/ip/pinmux/doc/_index.md)                                
        * [Pinmux Checklist](hw/ip/pinmux/doc/checklist.md)                                         
      * [Primitive Component: LFSR](hw/ip/prim/doc/prim_lfsr.md)                                    
      * [Primitive Component: Packer](hw/ip/prim/doc/prim_packer.md)                                
      * [Ibex RISC-V Core Wrapper Technical Specification](hw/ip/rv_core_ibex/doc/_index.md)        
        * [Ibex Processor Core Checklist](hw/ip/rv_core_ibex/doc/checklist.md)                      
      * [RISC-V Debug System Wrapper Technical Specification](hw/ip/rv_dm/doc/_index.md)            
      * [Interrupt Controller Technical Specification](hw/ip/rv_plic/doc/_index.md)                 
        * [RV_PLIC DV Plan](hw/ip/rv_plic/doc/dv_plan/index.md)                                     
      * [Timer HWIP Technical Specification](hw/ip/rv_timer/doc/_index.md)                          
        * [RV_TIMER Checklist](hw/ip/rv_timer/doc/checklist.md)                                     
          * [RV_TIMER DV Plan](hw/ip/rv_timer/doc/dv_plan/index.md)                                 
      * [SPI Device HWIP Technical Specification](hw/ip/spi_device/doc/_index.md)                   
        * [SPI_DEVICE Checklist](hw/ip/spi_device/doc/checklist.md)                                 
          * [SPI Device DV Plan](hw/ip/spi_device/doc/dv_plan/index.md)                             
      * [Bus Specification](hw/ip/tlul/doc/_index.md)                                               
        * [TL-UL Protocol Checker](hw/ip/tlul/doc/TlulProtocolChecker.md)                           
          * [TLUL XBAR DV Plan](hw/ip/tlul/doc/dv_plan/index.md)                                    
      * [UART HWIP Technical Specification](hw/ip/uart/doc/_index.md)                               
        * [UART Checklist](hw/ip/uart/doc/checklist.md)                                             
          * [UART DV Plan](hw/ip/uart/doc/dv_plan/index.md)                                         
      * [Simple USB Full Speed Device IP Technical Specification](hw/ip/usbdev/doc/_index.md)       
          * [USBDEV DV Plan](hw/ip/usbdev/doc/dv_plan/index.md)                                     
      * [USB UART HWIP Technical Specification](hw/ip/usbuart/doc/_index.md)                        
    * [Earl Grey Top Level Specification](hw/top_earlgrey/doc/_index.md)                            
    * [TOP Earl Grey](hw/top_earlgrey/dv/README.md)                                                 
      * [TL-UL Checklist](hw/top_earlgrey/ip/xbar/doc/checklist.md)                                 
    * [Git Considerations](hw/vendor/lowrisc_ibex/CONTRIBUTING.md)                                  
    * [Credits](hw/vendor/lowrisc_ibex/CREDITS.md)                                                  
    * [README.md](hw/vendor/lowrisc_ibex/README.md)                                                 
      * [Ibex simulation Control/Status Registers Testbench](hw/vendor/lowrisc_ibex/dv/cs_registers/README.md)
      * [Ibex simulation for RISC-V Compliance Testing](hw/vendor/lowrisc_ibex/dv/riscv_compliance/README.md)
      * [Ibex Simple System](hw/vendor/lowrisc_ibex/examples/simple_system/README.md)               
      * [How to Contribute](hw/vendor/lowrisc_ibex/vendor/google_riscv-dv/CONTRIBUTING.md)          
      * [Overview](hw/vendor/lowrisc_ibex/vendor/google_riscv-dv/HANDSHAKE.md)                      
      * [Overview](hw/vendor/lowrisc_ibex/vendor/google_riscv-dv/README.md)                         
      * [Verilator memory loading support](hw/vendor/lowrisc_ibex/dv/verilator/memutil/README.md)   
      * [Ibex RISC-V Core SoC Example](hw/vendor/lowrisc_ibex/examples/fpga/artya7-100/README.md)   
      * [Yosys/OpenSTA Ibex Synthesis Flow](hw/vendor/lowrisc_ibex/syn/README.md)                   
    * [Changelog](hw/vendor/pulp_riscv_dbg/CHANGELOG.md)                                            
    * [RISC-V Debug Support for PULP Cores](hw/vendor/pulp_riscv_dbg/README.md)                     
      * [Overview](hw/vendor/pulp_riscv_dbg/doc/debug-system.md)                                    
* [OpenTitan Software](sw/README.md)                                                                
  * [Work with software code in external repositories](sw/vendor/README.md)                         
    * [Boot ROM Overview](sw/device/boot_rom/README.md)                                             
    * [Software Tests Readme](sw/device/tests/index.md)                                             
      * [Building CoreMark](sw/device/benchmarks/coremark/README.md)                                
      * [Overview](sw/device/examples/hello_world/README.md)                                        
      * [`libbase`, the OpenTitan Standard Library](sw/device/lib/base/README.md)                   
      * [The OpenTitan DIF Library](sw/device/lib/dif/README.md)                                    
      * [`libruntime`, the OpenTitan Runtime Library](sw/device/lib/runtime/README.md)              
        * [OpenTitan Freestanding C Headers](sw/device/lib/base/freestanding/README.md)             
    * [SPI Flash](sw/host/spiflash/README.md)                                                       
    * [COREMARK® ACCEPTABLE USE AGREEMENT ](sw/vendor/eembc_coremark/LICENSE.md)                    
    * [Introduction](sw/vendor/eembc_coremark/README.md)                                            
      * [READM.md](sw/vendor/eembc_coremark/docs/READM.md)                                          
    * [Googletest Mocking (gMock) Framework](sw/vendor/google_googletest/googlemock/README.md)      
    * [How to become a contributor and submit your own code](sw/vendor/google_googletest/CONTRIBUTING.md)
    * [Google Test](sw/vendor/google_googletest/README.md)                                          
      * [Generic Build Instructions](sw/vendor/google_googletest/googletest/README.md)              
        * [gMock Cheat Sheet](sw/vendor/google_googletest/googlemock/docs/cheat_sheet.md)           
        * [gMock Cookbook](sw/vendor/google_googletest/googlemock/docs/cook_book.md)                
        * [gMock for Dummies {#GMockForDummies}](sw/vendor/google_googletest/googlemock/docs/for_dummies.md)
        * [Legacy gMock FAQ {#GMockFaq}](sw/vendor/google_googletest/googlemock/docs/gmock_faq.md)  
        * [The Problem](sw/vendor/google_googletest/googlemock/docs/pump_manual.md)                 
        * [Please Note:](sw/vendor/google_googletest/googlemock/scripts/README.md)                  
        * [Advanced googletest Topics](sw/vendor/google_googletest/googletest/docs/advanced.md)     
        * [Googletest FAQ](sw/vendor/google_googletest/googletest/docs/faq.md)                      
        * [Using GoogleTest from various build systems](sw/vendor/google_googletest/googletest/docs/pkgconfig.md)
        * [Googletest Primer](sw/vendor/google_googletest/googletest/docs/primer.md)                
        * [Googletest Samples {#samples}](sw/vendor/google_googletest/googletest/docs/samples.md)   
        * [Please Note:](sw/vendor/google_googletest/googletest/scripts/README.md)                  
    * [RISC-V Compliance Task Group](sw/vendor/riscv_compliance/README.md)                          
      * [RISC-V Targets](sw/vendor/riscv_compliance/riscv-target/README.md)                         
      * [RISC-V Test Suites](sw/vendor/riscv_compliance/riscv-test-suite/README.md)                 
        * [Overview](sw/vendor/riscv_compliance/riscv-target/opentitan/README.md)                   
* [OpenTitan Utilities](util/_index.md)                                                             
  * [Docker Container](util/container/README.md)                                                    
  * [FPGA Splice flow](util/fpga/README.md)                                                         
  * [Fpvgen: Initial FPV testbench generation tool](util/fpvgen/README.md)                          
  * [i2csvg -- Generate svg pictures of I2C commands from text description](util/i2csvg/README.md)  
    * [Using I2C master for SMBus commands](util/i2csvg/smbus/SMBus.md)                             
  * [Register generator `reggen` and `regtool`](util/reggen/README.md)                              
  * [Simple SPI Tests](util/simplespi/README.md)                                                    
  * [Testplanner tool](util/testplanner/README.md)                                                  
  * [TL-UL Crossbar RTL generator](util/tlgen/README.md)                                            
  * [Uvmdvgen: Initial testbench auto-generation tool](util/uvmdvgen/README.md)                     
  * [Wavegen -- Waveform generator in Python](util/wavegen/README.md)                               
    * [FOO DV plan](util/testplanner/examples/foo_dv_plan.md)

**Comment by imphil** [comment]
It's not uncommon to restrict the TOC to certain levels of depth, e.g. 2. That should keep things interesting for the reader.

**Comment by sjgitty** [comment]
> It's not uncommon to restrict the TOC to certain levels of depth, e.g. 2. That should keep things interesting for the reader.

Agreed except much of our interest stuff is buried deep beyond level 2,
and some of the level 1 and level 2 stuff is just filler. I guess the compromise
is that we force some catch-all documents of said "interesting stuff" at level
2 so that it is found in the TOC, the same way hw/_index.md has a map of
all of the hardware specifications.

**Comment by sjgitty** [comment]
Perhaps the to-do part of this issue is to create what the TOC-level-2 would look
like and suggest changes to our current organization where interesting content
would otherwise get lost.

**Comment by towoe** [comment]
Do we still want this?
With the "menu" now available on docs.opentitan.org I think it is better to have some hand crafted text guiding to reader to the important parts (if it would be the only content). But I can still try to get some automatically level 2 generated list which can then be added to the end for selected pages.

**Comment by imphil** [comment]
I'd either still go for it, or significantly reword it to not look like a TOC any more, but a summary of what's coming (as in https://opensocdebug.readthedocs.io/en/latest/). If it looks like a TOC, it should actually be one and not go out of date repeatedly.

**Comment by towoe** [comment]
Yes, we should not have a TOC which is not up-to-date.
Either none at all or auto generated.
The summary approach is also nice, should be possible with Hugo if the front matter of the documentation contains the description.
I will provide a PR with a suggestion for the automatic one and then we can decide which direction we will pursue.
